impl/src/main/java/org/jboss/weld/xml/BeansXmlValidator.java

```diff
@@ -0,0 +1,158 @@
+        if (beansXml == null) {
+            throw XmlLogger.LOG.loadError("unknown", null);
+        }
+        if (errorHandler == null) {
+            errorHandler = this;
+        }
+        Schema schema = cdi20Schema;
+
+        // First quick check of beans.xml to find out version
+        try (InputStream in = beansXml.openStream()) {
+            if (in.available() == 0) {
+                // The file is just acting as a marker file
+                return;
+            }
+            try (BufferedReader reader = new BufferedReader(new InputStreamReader(in, Charset.forName(StandardCharsets.UTF_8.name())))) {
+                String line;
+                while ((line = reader.readLine()) != null) {
+                    if (line.contains(XmlSchema.CDI11.getFileName())) {
+                        schema = cdi11Schema;
+                        break;
+                    } else if (line.contains(XmlSchema.CDI20.getFileName())) {
+                        break;
+                    }
+                }
+            }
+        } catch (IOException e) {
+            throw XmlLogger.LOG.loadError(beansXml, e);
+        }
+
+        if (schema == null) {
+            return;
+        }
+        Validator validator = schema.newValidator();
+        validator.setErrorHandler(errorHandler);
+        try {
+            validator.validate(new StreamSource(beansXml.toExternalForm()));
+        } catch (SAXException | IOException e) {
+            // No-op - validation is optional
+        }
+    }
+
+    public void validate(URL beansXml) {
+        validate(beansXml, this);
+    }
+
+    @Override
+    public void warning(SAXParseException e) throws SAXException {
+        XmlLogger.LOG.xsdValidationWarning(e.getSystemId(), e.getLineNumber(), e.getMessage());
+    }
+
+    @Override
+    public void error(SAXParseException e) throws SAXException {
+        if (e.getMessage().startsWith(VALIDATION_ERROR_CODE_CVC_ELT_1) && e.getMessage().contains(ROOT_ELEMENT_NAME)) {
+            // Ignore the errors we get when there is no schema defined
+            return;
+        }
+        XmlLogger.LOG.xsdValidationError(e.getSystemId(), e.getLineNumber(), e.getMessage());
+    }
+
+    @Override
+    public void fatalError(SAXParseException e) throws SAXException {
+        throw e;
+    }
+
+    private static Source[] loadXsds(XmlSchema[] schemas) {
+        List<Source> xsds = new ArrayList<>();
+        for (XmlSchema schema : schemas) {
+            Source source = loadXsd(schema.getFileName(), schema.getClassLoader());
+            if (source != null) {
+                xsds.add(source);
+            }
+        }
+        return xsds.toArray(EMPTY_SOURCE_ARRAY);
+    }
+
+    private static StreamSource loadXsd(String name, ClassLoader classLoader) {
+        InputStream in = classLoader.getResourceAsStream(name);
+        if (in == null) {
+            return null;
+        } else {
+            return new StreamSource(in);
```

Comment:
Maybe buffer this?

Reply:
XSDs are only loaded once (ideally). So I'd consider this a premature optimization ;-) The other issue is that we should probably close the stream...

Reply:
Up to you; I found during profiling that eliminating the many `read()` function calls in general was the best bang for the buck I could get.  XSDs like Java EE's tend to be huge, so….  But yes, I haven't profiled this.

I _believe_ that "sources" (like `StreamSource`) are guaranteed by contract to be closed for you on full consumption (just another bit of weirdness from the JAXP area).

Reply:
Also selfishly we don't have validation on anyway so for me personally it doesn't matter.

Reply:
Yes, at least the javadoc is lacking proper contract description :-(

Reply:
> Also selfishly...

Fair enough ;-). Validation errors result only in log warnings anyway...

Reply:
Ah; there's this: https://bugs.openjdk.java.net/browse/JDK-8139262

So probably want to pass in `InputStream`s explicitly to the `StreamSource` constructor wherever possible and then close them explicitly afterwards.  Bleagh.

Reply:
Hm, thanks for link!